executor/window.go

```diff
@@ -174,3 +175,130 @@ func (e *WindowExec) copyChk(chk *chunk.Chunk) {
+	// which belong to the same partition.
+	consumeGroupRows(ctx sessionctx.Context, rows []chunk.Row) ([]chunk.Row, error)
+	// appendResult2Chunk appends the final results to chunk.
+	// It is called when there are no more rows in current partition.
+	appendResult2Chunk(ctx sessionctx.Context, rows []chunk.Row, chk *chunk.Chunk, remained int) ([]chunk.Row, error)
+	// resetPartialResult resets the partial result to the original state for a specific window function.
+	resetPartialResult()
+}
+
+type aggWindowProcessor struct {
+	windowFunc    aggfuncs.AggFunc
+	partialResult aggfuncs.PartialResult
+}
+
+func (p *aggWindowProcessor) consumeGroupRows(ctx sessionctx.Context, rows []chunk.Row) ([]chunk.Row, error) {
+	err := p.windowFunc.UpdatePartialResult(ctx, rows, p.partialResult)
+	rows = rows[:0]
+	return rows, err
+}
+
+func (p *aggWindowProcessor) appendResult2Chunk(ctx sessionctx.Context, rows []chunk.Row, chk *chunk.Chunk, remained int) ([]chunk.Row, error) {
+	for remained > 0 {
+		// TODO: We can extend the agg func interface to avoid the `for` loop  here.
+		err := p.windowFunc.AppendFinalResult2Chunk(ctx, p.partialResult, chk)
+		if err != nil {
+			return rows, err
+		}
+		remained--
+	}
+	return rows, nil
+}
+
+func (p *aggWindowProcessor) resetPartialResult() {
+	p.windowFunc.ResetPartialResult(p.partialResult)
+}
+
+type rowFrameWindowProcessor struct {
+	windowFunc    aggfuncs.AggFunc
+	partialResult aggfuncs.PartialResult
+	start         *core.FrameBound
+	end           *core.FrameBound
+	curRowIdx     uint64
+}
+
+func (p *rowFrameWindowProcessor) getStartOffset(numRows uint64) uint64 {
+	if p.start.UnBounded {
+		return 0
+	}
+	switch p.start.Type {
+	case ast.Preceding:
+		if p.curRowIdx >= p.start.Num {
+			return p.curRowIdx - p.start.Num
+		}
+		return 0
+	case ast.Following:
+		offset := p.curRowIdx + p.start.Num
+		if offset >= numRows {
+			return numRows
+		}
+		return offset
+	case ast.CurrentRow:
+		return p.curRowIdx
+	}
+	// It will never reach here.
+	return 0
+}
+
+func (p *rowFrameWindowProcessor) getEndOffset(numRows uint64) uint64 {
+	if p.end.UnBounded {
+		return numRows
+	}
+	switch p.end.Type {
+	case ast.Preceding:
+		if p.curRowIdx >= p.end.Num {
+			return p.curRowIdx - p.end.Num + 1
+		}
+		return 0
+	case ast.Following:
+		offset := p.curRowIdx + p.end.Num
+		if offset >= numRows {
+			return numRows
+		}
+		return offset + 1
+	case ast.CurrentRow:
+		return p.curRowIdx + 1
+	}
+	// It will never reach here.
+	return 0
+}
+
+func (p *rowFrameWindowProcessor) consumeGroupRows(ctx sessionctx.Context, rows []chunk.Row) ([]chunk.Row, error) {
+	return rows, nil
+}
+
+func (p *rowFrameWindowProcessor) appendResult2Chunk(ctx sessionctx.Context, rows []chunk.Row, chk *chunk.Chunk, remained int) ([]chunk.Row, error) {
+	numRows := uint64(len(rows))
+	for remained > 0 {
+		start := p.getStartOffset(numRows)
+		end := p.getEndOffset(numRows)
+		var err error
+		if start >= end {
```

Comment:
What about doing this by sliding window algorithm?

Reply:
It is left as further optimizations.

Reply:
How about leave a `TODO` comment for this.

Reply:
Done.